### PR TITLE
Add a suite of external integration tests

### PR DESCRIPTION
**Pull request by akshayjshah**
Before making any significant changes to the config logic, add a suite
of unit tests that exercise the externally-visible behavior of the
current implementation. A number of these tests currently fail, so
they're skipped and marked with todos.

This includes tests that replicate the following issues:

- https://github.com/uber-go/config/issues/72
- https://github.com/uber-go/config/issues/59
- https://github.com/uber-go/config/issues/76
- https://github.com/uber-go/config/issues/75

### CODE REVIEW

File: config_ext_test.go [hunk 1]
```diff
@@ -0,0 +1,350 @@
...
+			Value:     val,
+			HasValue:  true,
+			Interface: "bottom",
+			Populate:  &s,
+		})
+		assert.Equal(t, "bottom", s, "unexpected result after populate")
+	})
+
+	t.Run("deep copy", func(t *testing.T) {
+		t.Skip("TODO: mutations are visible to other callers")
```
Comment (akshayjshah) [inline config_ext_test.go]: This is probably the most serious of the existing bugs.

File: config_ext_test.go [hunk 2]
```diff
@@ -0,0 +1,350 @@
...
+		run(TestCase{
+			Value:     provider.Get("occupants").Get("honda").Get("driver"),
+			HasValue:  true,
+			Interface: "jane",
+			Populate:  &s,
+		})
+		assert.Equal(t, "jane", s, "unexpected result after populate")
+	})
+
+	t.Run("default missing value", func(t *testing.T) {
```
Comment (amckinney) [inline config_ext_test.go]: Is this test identical to the above?

Comment (amckinney) [reply config_ext_test.go]: Based on the test below, I have a feeling this is supposed to test a call to `WithDefault` where the result would have otherwise been `nil`.

File: config_ext_test.go [hunk 3]
```diff
@@ -0,0 +1,350 @@
...
+		})
+		assert.Equal(t, map[string]string{
+			"honda":  "civic",
+			"toyota": "camry",
+			"nissan": "altima",
+			"ford":   "fiesta",
+		}, m, "unexpected result after populate")
+	})
+
+	t.Run("default replaces sequences", func(t *testing.T) {
```
Comment (amckinney) [inline config_ext_test.go]: The test description is a little misleading - the default doesn't replace the configured sequence in this case.

File: config_ext_test.go [hunk 4]
```diff
@@ -0,0 +1,350 @@
...
+	t.Run("default ignores type mismatches", func(t *testing.T) {
+		// TODO: We should return an error in strict mode.
+		// fun is a sequence, so we shouldn't be able to merge in a mapping
+		val, err := provider.Get("fun").WithDefault(map[string]string{"dodge": "viper"})
+		assert.NoError(t, err, "error on type mismatch in permissive mode")
+		assert.Equal(t, []interface{}{"maserati", "lamborghini"}, val.Value(), "wrong merged output")
+	})
+
+	t.Run("chained defaults", func(t *testing.T) {
+		// Defaults should be FIFO: once we've set a default value, we can't later
```
Comment (amckinney) [inline config_ext_test.go]: nit: not sure if FIFO is the best description here since that might imply the next-queued default would be used on a subsequent call to `Get` (i.e. `boom` would be used after `bottom`). Seems strange to use "idempotent" in this context, but something similar to "no effect after the first call", might be more appropriate.

File: config_ext_test.go (hunk 1 above)
Comment (amckinney) [reply config_ext_test.go]: ...oh no

**Comment by amckinney** [comment]
Some minor comments, but otherwise looks great!

File: config_ext_test.go (hunk 2 above)
Comment (akshayjshah) [reply config_ext_test.go]: :man_facepalming: you're correct.

File: config_ext_test.go (hunk 3 above)
Comment (akshayjshah) [reply config_ext_test.go]: Good point, I'll change - how does "default sequences aren't merged" sound?

File: config_ext_test.go (hunk 4 above)
Comment (akshayjshah) [reply config_ext_test.go]: Also a good point, I'll reword this.

**Comment by akshayjshah** [comment]
@SyrieBianco this is the config package we were talking about.

File: config_ext_test.go [hunk 5]
```diff
@@ -0,0 +1,350 @@
...
+	})
+
+	t.Run("ignored field", func(t *testing.T) {
+		type cfg struct {
+			Boring []string `yaml:"-"`
+		}
+
+		var c cfg
+		require.NoError(t, provider.Get(Root).Populate(&c))
+		assert.Equal(t, 0, len(c.Boring), "wrong number of fun cars")
```
Comment (prashantv) [inline config_ext_test.go]: optionally: `assert.Empty`

File: config_ext_test.go [hunk 6]
```diff
@@ -0,0 +1,350 @@
...
+  - porsche
+practical:
+  toyota: camry
+  honda: accord
+occupants:
+  honda:
+    driver: jane
+    backseat: [nate]
+`)
+	second := strings.NewReader(`
```
Comment (prashantv) [inline config_ext_test.go]: can we also add a test where a value in `first` is set to `~` in `second`, ensuring that overrides to `nil` work

File: config_ext_test.go [hunk 7]
```diff
@@ -0,0 +1,350 @@
...
+			HasValue:  true,
+			Interface: nil,
+			Populate:  &s,
+		})
+		assert.Equal(t, "", s, "unexpected result after populate")
+	})
+
+	t.Run("default merges maps", func(t *testing.T) {
+		var m map[string]string
+		val, err := provider.Get("practical").WithDefault(map[string]string{"ford": "fiesta"})
```
Comment (prashantv) [inline config_ext_test.go]: would also add a key that exists in `first` to make sure that the merge replaces existing keys from defaults

File: config_ext_test.go [hunk 8]
```diff
@@ -0,0 +1,350 @@
...
+			"nissan": "altima",
+		}, m, "unexpected result after populate")
+	})
+
+	t.Run("slice", func(t *testing.T) {
+		var s []string
+		run(TestCase{
+			Value:     provider.Get("fun"),
+			HasValue:  true,
+			Interface: []interface{}{"maserati", "lamborghini"},
```
Comment (prashantv) [inline config_ext_test.go]: for some reason, i thought that the custom merging logic merged lists, rather than replacing, but i guess it matches the yaml unmarshaler?

File: config_ext_test.go [hunk 9]
```diff
@@ -0,0 +1,350 @@
...
+	"testing"
+
+	. "go.uber.org/config"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestIntegration(t *testing.T) {
+	first := strings.NewReader(`
```
Comment (prashantv) [inline config_ext_test.go]: is it worth using the yaml pointers to make sure they work as expected?

File: config_ext_test.go (hunk 3 above)
Comment (amckinney) [reply config_ext_test.go]: 👍

File: config_ext_test.go (hunk 5 above)
Comment (akshayjshah) [reply config_ext_test.go]: For no particular reason, I usually avoid that helper. Alex mentioned the same thing, so I'll change.

File: config_ext_test.go (hunk 9 above)
Comment (akshayjshah) [reply config_ext_test.go]: Good point, definitely worthwhile.

File: config_ext_test.go (hunk 8 above)
Comment (akshayjshah) [reply config_ext_test.go]: Yep. There are a bunch of internal tickets where we discussed merging lists, but the actual implementation appears to match the yaml package pretty closely.

File: config_ext_test.go (hunk 6 above)
Comment (akshayjshah) [reply config_ext_test.go]: Done.

File: config_ext_test.go (hunk 7 above)
Comment (akshayjshah) [reply config_ext_test.go]: *Great* call. As it turns out, that test fails...so the current behavior of `WithDefault` is to merge in new keys, but not replace existing keys. I'm calling this a bug rather than a behavior to preserve.

**Comment by akshayjshah** [comment]
Pushed another commit addressing CR feedback. I've included tests for errors later in the stack...but they're definitely coming.

File: config_ext_test.go [hunk 10]
```diff
@@ -31,6 +35,7 @@ fun:
 practical:
   honda: civic
   nissan: altima
+antique: model_t
```
[the remark this answers is not in the document]
Comment (akshayjshah) [reply config_ext_test.go]: #72 is covered explicitly below, applying a non-nil default to an explicitly nulled-out value. Can easily switch this around too.

File: config_ext_test.go [hunk 11]
```diff
@@ -0,0 +1,426 @@
...
+		})
+		assert.Equal(t, Occupants{
+			Driver:    "jane",
+			Passenger: "arthur",
+			Backseat:  []string{"nora"},
+		}, o, "unexpected result after populate")
+	})
+
+	t.Run("nil merged into scalar", func(t *testing.T) {
+		t.Skip("TODO: ignores explicit nil")
```
Comment (prashantv) [inline config_ext_test.go]: thanks for adding this -- one other case that might be worth adding is nil merging into an existing sequence/map. I'm not sure of the right behaviour, I'd assume replace in both cases, but since maps merge, perhaps it just gets ignored?

Comment (akshayjshah) [reply config_ext_test.go]: Good point. I'd expect sequences to be replaced and maps to be merged, since that's the behavior for non-nil things.

File: config_ext_test.go [hunk 12]
```diff
@@ -0,0 +1,461 @@
...
+		// lower-priority values unchanged.
+		var m map[string]string
+		run(t, TestCase{
+			Value:     provider.Get("antique_mapping"),
+			HasValue:  true,
+			Interface: map[interface{}]interface{}{"ford": "model_t"},
+			Populate:  &m,
+		})
+		assert.Equal(t, map[string]string{"ford": "model_t"}, m, "unexpected result after Populate")
+	})
```
Comment (akshayjshah) [inline config_ext_test.go]: In Go, this is consistent with the rest of the merging behavior - maps are deep-merged, and `nil` is a valid map with no keys.

However, it's a little weird. I'm open to the idea that we treat nils and empty maps differently here - `nil` replaces the existing value, and `{}` has no effect.

Comment (prashantv) [reply config_ext_test.go]: I'd prefer that behaviour, but we should probably look at what the YAML library does in this case. We may want to match it (since the rest of the library seems to try and match it)?

File: config_ext_test.go [hunk 13]
```diff
@@ -0,0 +1,461 @@
...
+		assert.Equal(t, []string{"maserati", "lamborghini"}, s, "unexpected result after populate")
+	})
+
+	t.Run("default ignores type mismatches", func(t *testing.T) {
+		// TODO: We should return an error in strict mode.
+		// fun is a sequence, so we shouldn't be able to merge in a mapping
+		val, err := provider.Get("fun").WithDefault(map[string]string{"dodge": "viper"})
+		assert.NoError(t, err, "error on type mismatch in permissive mode")
+		assert.Equal(t, []interface{}{"maserati", "lamborghini"}, val.Value(), "wrong merged output")
+	})
```
Comment (akshayjshah) [inline config_ext_test.go]: This is wrong; defaults should act like a `base_base.yaml`, which controls the type of each key. The end result here should be `{dodge: viper}`.

File: config_ext_test.go [hunk 14]
```diff
@@ -0,0 +1,461 @@
...
+
+	t.Run("chained defaults", func(t *testing.T) {
+		// Once we've set a default, the key has a value. Subsequent calls to
+		// WithDefault have no effect.
+		val, err := provider.Get("top").WithDefault(map[string]string{"middle": "bottom"})
+		require.NoError(t, err, "couldn't set first default")
+		val, err = val.Get("middle").WithDefault("boom")
```
Comment (akshayjshah) [inline config_ext_test.go]: Should be same type as first `WithDefault`.

For consistency, we should also be treating *each* `WithDefault` as though it were the base for a merge. Here, we should be able to set `{other_middle: other_bottom}` in a second call to `WithDefault` and see both keys and values in the final result.

File: config_ext_test.go (hunk 12 above)
Comment (akshayjshah) [reply config_ext_test.go]: Yeah - after talking through this with Alex, we came to the same conclusion. That's also how the upstream YAML lib behaves, so we're safe.
